### DISCUSSION TITLE
Compression level (of tiff files)

### DISCUSSION THREAD

**Discussion by GiottoVerducci** [ImageMagick/ImageMagick#1901] [General]
I have large tiff files (often using LZW compression) that I would like to recompress. I know that one can use the -compress ZIP option (http://www.imagemagick.org/discourse-server/viewtopic.php?t=33141) but it will result in larger files than if I opened it with Photoshop and saved it with the ZIP compression.

For instance:

LZW: 160,091,372 bytes / Endianess: LSB
ZIP using Photoshop: 59,463,688 bytes / Endianess: LSB
ZIP using ImageMagick: 61,148,606 / Endianess: MSB
ZIP using ImageMagick -quality 90: 59,535,052 / Endianess: MSB ()

(I mention endianess because it could have a slight influence on compression -- that's highly hypothetical)

Here's the original file "identify" informations:

```
Image: fr_01c.tif
  Format: TIFF (Tagged Image File Format)
  Mime type: image/tiff
  Class: DirectClass
  Geometry: 12241x11662+0+0
  Resolution: 1470x1470
  Print size: 8.32721x7.93333
  Units: PixelsPerInch
  Colorspace: sRGB
  Type: TrueColor
  Endianess: LSB
  Depth: 8-bit
  Channel depth:
    Red: 8-bit
    Green: 8-bit
    Blue: 8-bit
  Channel statistics:
    Pixels: 142754542
    Red:
      min: 0  (0)
      max: 255 (1)
      mean: 54.3054 (0.212962)
      standard deviation: 56.8854 (0.22308)
      kurtosis: 3.27646
      skewness: 1.67672
      entropy: 0.753559
    Green:
      min: 0  (0)
      max: 255 (1)
      mean: 28.3997 (0.111371)
      standard deviation: 48.2422 (0.189185)
      kurtosis: 14.0485
      skewness: 3.6991
      entropy: 0.638771
    Blue:
      min: 0  (0)
      max: 255 (1)
      mean: 33.8041 (0.132565)
      standard deviation: 50.5792 (0.19835)
      kurtosis: 10.6819
      skewness: 3.25092
      entropy: 0.664939
  Image statistics:
    Overall:
      min: 0  (0)
      max: 255 (1)
      mean: 38.8364 (0.1523)
      standard deviation: 51.9023 (0.203538)
      kurtosis: 7.28985
      skewness: 2.623
      entropy: 0.685756
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 12241x11662+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: LZW
  Orientation: TopLeft
  Profiles:
    Profile-icc: 3144 bytes
  Properties:
    date:create: 2020-04-16T08:24:15+00:00
    date:modify: 2020-01-12T14:10:41+00:00
    icc:copyright: Copyright (c) 1998 Hewlett-Packard Company
    icc:description: sRGB IEC61966-2.1
    icc:manufacturer: IEC http://www.iec.ch
    icc:model: IEC 61966-2.1 Default RGB colour space - sRGB
    signature: 2bfac7bd58af15fe3892c5a53e5f2eb00c59ee3aa6869c5dfb806907b2f0622e
    tiff:alpha: unspecified
    tiff:endian: lsb
    tiff:photometric: RGB
    tiff:rows-per-strip: 1
  Artifacts:
    verbose: true
  Tainted: False
  Filesize: 152.675MiB
  Number pixels: 142.755M
  Pixels per second: 70.9546MP
  User time: 2.016u
  Elapsed time: 0:03.011
  Version: ImageMagick 7.0.10-2 Q16 x86 2020-03-22 http://www.imagemagick.org
```

I was wondering if there was a way to achieve a higher compression level:
 - using a parameter I couldn't find (better than -quality 90)
 - recompiling ImageMagick with a different compression algorithm (though it's probably too time-consuming)

**Comment by GiottoVerducci** [comment]
I just found some "better" deflate algorithm are freely available, for instance: 
https://en.wikipedia.org/wiki/AdvanceCOMP
https://github.com/amadvance/advancecomp/tree/master/libdeflate
or https://github.com/google/zopfli

Maybe this could help?

**Comment by GiottoVerducci** [comment]
I confirm that I was able to write a POC that takes the tiff file, extract the content of the first "strip" (see https://www.itu.int/itudoc/itu-t/com16/tiff-fx/docs/tiff6.pdf and https://www.alternatiff.com/resources/TIFFphotoshop.pdf), decompress it, recompress it with a higher compression level (using the first c# lib I could find, that is https://www.nuget.org/packages/DotNetZip -- zopfli should compress even better), and update the strip with a smaller content without altering the image.

**Comment by GiottoVerducci** [comment]
I found that with the parameter "-quality 90", you can achieve a better compression but it's quite long, and still bigger than the file produced by Photoshop:
59,535,052 bytes

**Comment by urban-warrior** [comment]
ImageMagick leverages the libtiff delegate library to compress the image into the TIFF image format.  Any suggestions you have for improving compression should likely be posted to the libtiff developers.  Once a compression improvement makes it way to the official libtiff distribution, ImageMagick will inherit the benefit when writing TIFF images.

**Reply by GiottoVerducci** [reply]
I haven't been working on the subject for a long time, but eventually I contacted the libtiff team. (https://gitlab.com/libtiff/libtiff/-/issues/446) It appears that they support the use of the libDeflate library and compression levels up to 12. Is ImageMagick using that feature?  I suppose not, since I the quality ranges from 0 to 100, and "the quality value sets the zlib compression level (quality / 10) [for the MNG and PNG image formats]" (and tiff I suppose).
Also, the size difference between images produced by ImageMagick and Photoshop make me think the best compression is not used.

**Reply by pbl987** [reply]
Thank you for your effort, i hope ImageMagick will get an upgrade soon.
I am looking for this too.

**Comment by kmilos** [comment]
You will also find that playing with the `tiff:rows-per-strip` [option ](https://imagemagick.org/script/defines.php)could help slightly with the compression ratio, but ensuring the horizontal `tiff:predictor` is enabled makes the larger impact (I suspect that should be already the case given the huge difference between the original LZW file and all the ZIPs...).
